File: src/patternView.cxx
/***************************************************************
*
* (c) 2009-2010 Markus Dittrich 
*
* This program is free software; you can redistribute it 
* and/or modify it under the terms of the GNU General Public 
* License Version 3 as published by the Free Software Foundation. 
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License Version 3 for more details.
*
* You should have received a copy of the GNU General Public 
* License along with this program; if not, write to the Free 
* Software Foundation, Inc., 59 Temple Place - Suite 330, 
* Boston, MA 02111-1307, USA.
*
****************************************************************/

/** Qt headers */
#include <QDebug>
#include <QMouseEvent>
#include <QRubberBand>
#include <QGraphicsSceneMouseEvent>
#include <QWheelEvent>

#include <QScrollBar>

/** local headers */
#include "basicDefs.h"
#include "graphicsScene.h"
#include "patternGridItem.h"
#include "patternView.h"


QT_BEGIN_NAMESPACE


/**************************************************************
 *
 * PUBLIC FUNCTIONS 
 *
 **************************************************************/

//-------------------------------------------------------------
// constructor
//-------------------------------------------------------------
PatternView::PatternView(GraphicsScene* aScene, QWidget* myParent)
  :
  QGraphicsView(aScene, myParent),
  canvas_(aScene),
  rubberBandOn_(false)
{
  status_ = SUCCESSFULLY_CONSTRUCTED;
}


//--------------------------------------------------------------
// main initialization routine
//--------------------------------------------------------------
bool PatternView::Init()
{
  if ( status_ != SUCCESSFULLY_CONSTRUCTED )
  {
    return false;
  }

  /* set some basic properties */
  setDragMode(QGraphicsView::NoDrag);
  setRenderHints(QPainter::Antialiasing);

  initialize_rubberband_();
  visible_in_view();
  
  return true;
}


/**************************************************************
 *
 * PUBLIC FUNCTIONS
 *
 *************************************************************/
 
/**************************************************************
 *
 * PUBLIC SLOTS
 *
 *************************************************************/

//-------------------------------------------------------------
// slot responsible for making sure all graphics items
// that should be are accesible in the view
//-------------------------------------------------------------
void PatternView::accessible_in_view()
{
  // fix our view on the initial grid size; otherwise items
  // added to the legend (even if inivisible) will cause
  // spurious re-scaling
  QRectF myScene = canvas_->get_visible_area();
  myScene.adjust(-50,-50,100,100);
  setSceneRect(myScene);

  // FIXME: this is a dirty hack; there has to be a better
  // way. The following is happening: If we are in a view with
  // no scrollbar and the above call to setSceneRect forces
  // scrollbars to appear due to the scene growing, Qt will
  // move the scrollbars such that upper left corner of the
  // default view is (0,0) instead of the true upper left 
  // corner of the view. Note that all is well if we already
  // have scrollbars that are set at a given value. Hopefully,
  // the below hack won't mess with these.
  int horizontalScrollMin = horizontalScrollBar()->minimum();
  int horizontalScrollVal = horizontalScrollBar()->value();

  if ((horizontalScrollMin <= 0) && (horizontalScrollVal == 0))
  {
    horizontalScrollBar()->setValue(horizontalScrollMin);
  }

  int verticalScrollMin = verticalScrollBar()->minimum();
  int verticalScrollVal = verticalScrollBar()->value();

  if ((verticalScrollMin <= 0) && (verticalScrollVal == 0))
  {
    verticalScrollBar()->setValue(verticalScrollMin);
  }
}



//-------------------------------------------------------------
// slot responsible for making sure all graphics items
// that should be are visible in the view
//-------------------------------------------------------------
void PatternView::visible_in_view()
{
  accessible_in_view();

  /* center the view on the pattern grid */
  QPoint gridCenter = canvas_->get_grid_center();
  centerOn(mapFromScene(gridCenter));

  setMatrix(QMatrix());
}



//-------------------------------------------------------------
// slot responsible for zooming into the canvas
//-------------------------------------------------------------
void PatternView::zoom_in()
{
  QPointF center(mapToScene(rect()).boundingRect().center());
  scale(1.1,1.1);
  centerOn(center);
}


//-------------------------------------------------------------
// slot responsible for zooming out of the canvas
//-------------------------------------------------------------
void PatternView::zoom_out()
{
  QPointF center(mapToScene(rect()).boundingRect().center());
  scale(0.9,0.9);
  centerOn(center);
}



//------------------------------------------------------------
// SLOTS for paning
//------------------------------------------------------------
void PatternView::pan_down()
{
  translate(0,-30);
}


void PatternView::pan_left()
{
  translate(30,0);
}


void PatternView::pan_right()
{
  translate(-30,0);
}


void PatternView::pan_up()
{
  translate(0,30);
}



/**************************************************************
 *
 * PROTECTED 
 *
 *************************************************************/

//-------------------------------------------------------------
// deal with mouse release events and rubber band 
// events if requested
//-------------------------------------------------------------
void PatternView::mouseReleaseEvent(QMouseEvent* evt)
{
  if (rubberBandOn_)
  {
    /* retrieve final rubberBand geometry and pick
     * and select them */
    QRectF finalGeometry(
        mapToScene(rubberBand_->geometry()).boundingRect());
    canvas_->select_region(finalGeometry);

    rubberBandOn_ = false;
    rubberBand_->hide();
  }

  QGraphicsView::mouseReleaseEvent(evt);
}


//-------------------------------------------------------------
// deal with mouse press events and rubber band 
// events if requested
//-------------------------------------------------------------
void PatternView::mousePressEvent(QMouseEvent* evt)
{
  if (evt->modifiers().testFlag(Qt::ControlModifier))
  {
    rubberBandOn_ = true;
    rubberBandOrigin_ = evt->pos();
    rubberBand_->move(evt->pos());
    rubberBand_->resize(0,0);
    rubberBand_->show();
  }
  
  QGraphicsView::mousePressEvent(evt);
}


//-------------------------------------------------------------
// deal with mouse move events and rubber band 
// geometry changes 
//-------------------------------------------------------------
void PatternView::mouseMoveEvent(QMouseEvent* evt)
{
  if (rubberBandOn_)
  {
    rubberBand_->setGeometry(QRect(rubberBandOrigin_,
                                   evt->pos()).normalized());
  }
    
  QGraphicsView::mouseMoveEvent(evt);
}


//--------------------------------------------------------------
// event handler for mouse wheel events
//--------------------------------------------------------------
void PatternView::wheelEvent(QWheelEvent* aWheelEvent)
{
  if (aWheelEvent->modifiers().testFlag(Qt::ControlModifier) 
      && aWheelEvent->delta() > 0)
  {
    zoom_in();
  }
  else if (aWheelEvent->modifiers().testFlag(Qt::ControlModifier)
           && aWheelEvent->delta() < 0)
  {
    zoom_out();
  }
  else
  {
    QGraphicsView::wheelEvent(aWheelEvent);
  }
}


/**************************************************************
 *
 * PRIVATE SLOTS
 *
 *************************************************************/

/*************************************************************
 *
 * PRIVATE MEMBER FUNCTIONS
 *
 *************************************************************/

//-------------------------------------------------------------
// set up the rubber band
//-------------------------------------------------------------
void PatternView::initialize_rubberband_()
{
  rubberBand_ = new QRubberBand(QRubberBand::Rectangle, this);

  /* change colors
   * FIXME: Background coloring does not work for some reason */
  rubberBand_->setAutoFillBackground(true);
  rubberBand_->setWindowOpacity(0.4);
  QPalette aPalette = rubberBand_->palette();
  aPalette.setBrush(QPalette::Base, Qt::blue);
  aPalette.setBrush(QPalette::WindowText, Qt::red);
  rubberBand_->setPalette(aPalette);
}


QT_END_NAMESPACE

File: sconcho/gui/pattern_view.py
# -*- coding: utf-8 -*-
######################################################################## #
# (c) 2009-2011 Markus Dittrich
#
# This program is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public
# License Version 3 as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License Version 3 for more details.
#
# You should have received a copy of the GNU General Public
# License along with this program; if not, write to the Free
# Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.
#
#######################################################################


from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals
from __future__ import absolute_import

import operator
from PyQt4.QtCore import (Qt, QRect, QSize, QPointF, QSizeF,
                          QRectF) 
from PyQt4.QtGui import (QGraphicsView, QRubberBand, QPainter)

from sconcho.util.canvas import visible_bounding_rect


#########################################################
## 
## class for managing the actual pattern canvas
##
#########################################################
class PatternView(QGraphicsView):


    def __init__(self, parent = None):

        super(PatternView, self).__init__(parent)

        self.setRenderHint(QPainter.Antialiasing)
        self.setRenderHint(QPainter.TextAntialiasing)
        self.setCacheMode(QGraphicsView.CacheBackground)
        self.setResizeAnchor(QGraphicsView.AnchorViewCenter)

        # initialize the rubberBand
        self.rubberBand = QRubberBand(QRubberBand.Rectangle, self)
        self.rubberBand.hide()
        self.rubberBandOrigin = None



    def mousePressEvent(self, event):
        """ Handles mouse press events.

        We override this function to check if the user requested
        a rubberBand selection.
        NOTE: We only propagate the event if no rubberBand was selected
        to avoid selection of the pattern grid item where the click
        occured.
        """

        if event.modifiers() & Qt.ShiftModifier:
            self.rubberBandOrigin = event.pos()
            self.rubberBand.setGeometry(QRect(self.rubberBandOrigin, 
                                              QSize()))
            self.rubberBand.show()
        
        return QGraphicsView.mousePressEvent(self, event)



    def mouseMoveEvent(self, event):
        """ Handle mouse move events.

        If a rubberBand is active we adjust the size otherwise
        we just hand off the signal.
        """

        if self.rubberBandOrigin:
            self.rubberBand.setGeometry(QRect(self.rubberBandOrigin,
                                              event.pos()).normalized())

        return QGraphicsView.mouseMoveEvent(self, event)
        


    def mouseReleaseEvent(self, event):
        """ Handle mouse release events.

        If a rubberBand is active we tell our GraphicsScene what are
        was selected and then disable the rubberBand.
        """
        
        if self.rubberBandOrigin:
            self.scene().select_region((self.mapToScene(
                self.rubberBand.geometry())))
            self.rubberBandOrigin = None
            self.rubberBand.hide()
        
        return QGraphicsView.mouseReleaseEvent(self, event)


    
    def wheelEvent(self, event):
        """ Mouse wheel events cause zooming in and out. """

        if event.modifiers() & Qt.ControlModifier:
            if event.delta() > 0:
                self.zoom_in()
            else:
                self.zoom_out()
        else:
            QGraphicsView.wheelEvent(self, event)



    def zoom_in(self):
        """ Zoom in by 10% """

        self.scale(1.1, 1.1)



    def zoom_out(self):
        """ Zoom out by 10% """

        self.scale(0.9, 0.9)



    def fit_scene(self):
        """ Fit scene into canvas. """
        
        margin = 50.0
        rawBoundary = visible_bounding_rect(self.scene().items())
        rawBoundary.adjust(-margin, -margin, margin, margin)
        self.fitInView(rawBoundary, Qt.KeepAspectRatio)



    def normal_view(self):
        """ Resets scene to normal (initial) view. """

        self.resetMatrix()
